### PR TITLE
GTEST/UCP: Use perf retry count for WFE measurements

### LINKED ISSUE #5913: failure in shm/test_ucp_wfe.envelope/0 <shm,> on swx-rdmz-ucx-new-02
### Describe the bug
happened here https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/10884/logs/135
```
2020-11-16T15:18:17.0754379Z [----------] 1 test from shm/test_ucp_wfe
2020-11-16T15:18:17.0755078Z [ RUN      ] shm/test_ucp_wfe.envelope/0 <shm,>
2020-11-16T15:18:17.3383705Z [     INFO ]                put latency : 3.415 usec (performance not checked)
2020-11-16T15:18:17.6340637Z [     INFO ]       put latency with WFE : 0.129 usec
2020-11-16T15:18:19.9261461Z [     INFO ]       put latency with WFE : 0.132 usec (attempt 1)
2020-11-16T15:18:22.2307138Z [     INFO ]       put latency with WFE : 0.131 usec (attempt 2)
2020-11-16T15:18:24.5141819Z [     INFO ]       put latency with WFE : 0.132 usec (attempt 3)
2020-11-16T15:18:26.5167379Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test_perf.cc:308: Failure
2020-11-16T15:18:26.5169207Z Failed
2020-11-16T15:18:26.5170347Z Invalid put latency with WFE performance, expected: 2.39..8.54
2020-11-16T15:18:26.5177509Z [  FAILED  ] shm/test_ucp_wfe.envelope/0, where GetParam() = shm, (9442 ms)
2020-11-16T15:18:26.5179259Z [----------] 1 test from shm/test_ucp_wfe (9442 ms total)
2020-11-16T15:18:26.5179931Z 
```

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Use perf retry count for WFE measurements.

## Why ?

Fixes #5913
No need to overwrite the value, since test_jenkins script uses `10` and could be adjusted by test system specifically for setups in the future.
User has to specify `-p <unsigned interger>` option to `gtest`

## How ?

1. Fix code duplication for adjusting tests on ARM by defining `test_adjust()` method `perf_test` class. Use it in UCP and UCT.
2. Remove overwriting perf count value in WFE test.
3. Minor code cleanups.

### CODE REVIEW

**Comment by shamisp** [comment]
Let me give it a try. I don't think this solves #5913. The problem is that the fist performance collecting (reference) iteration gets some very noisy result (probably the machine is heavily oversubscribed or swapped). The comparison iteration gets better results but since we check the low boundary it fails. My current idea is to introduce additional flags for low boundary and high boundary. Then explicitly ask the test not to check the low boundary - aka if you get latency that is lower than your reference run, it is PASS.

**Comment by dmitrygx** [comment]
@shamisp I see
consider this PR as aligning WFE perf with other perf tests (UCP/UCT). we should use `-p` instead of adding its own value

**Review by shamisp — APPROVED** [review]

**Comment by shamisp** [comment]
@dmitrygx - sure, I have no problem with this.  I just highlighted that this does not really solve the original issue.

**Comment by dmitrygx** [comment]
> @dmitrygx - sure, I have no problem with this. I just highlighted that this does not really solve the original issue.

yes, I see
thank you for the explanation!

**Comment by dmitrygx** [comment]
@brminich @yosefe could you review pls?

**Comment by yosefe** [comment]
we should put WFE tests in same test fixture and file as ucp perf test, it will eliminate most code duplication
@shamisp when do you plan to fix it?
@dmitrygx as intermediate solution, can we just ignore WFE failures for gtest (so it will not fail CI)?

**Comment by dmitrygx** [comment]
> @dmitrygx as intermediate solution, can we just ignore WFE failures for gtest (so it will not fail CI)?

yes, we can
we could skip in case of unmatched perf

**Comment by shamisp** [comment]
@yosefe I'm working on it. I would keep it as a separate because it serves different goal and the code duplication is minimal.

**Comment by shamisp** [comment]
@dmitrygx you can pull my changed from my branch here https://github.com/shamisp/ucx/tree/topic/ucp/gtest_wfe_ps
it will address this immediate issue without "skip". It is based of your branch so you can just pull the changes into this PR.

**Comment by shamisp** [comment]
@dmitrygx Addressed your comments and just pushed changes addressing @yosefe comments in #5861.

**Comment by dmitrygx** [comment]
> addressing @yosefe comments in #5861.

@shamisp looks like, still using MAX_ITER instead of `ucs::perf_retry_count`. I think better to use `ucs_max(1, ucs::perf_retry_count)`, since it could be configured by a user

also if i'm not mistaken @yosefe asked to move wait_mem testing to `gtest/ucp/test_ucp_perf.cc` file

>we should put WFE tests in same test fixture and file as ucp perf test, it will eliminate most code duplication

**Comment by shamisp** [comment]
> > addressing @yosefe comments in #5861.
> 
> @shamisp looks like, still using MAX_ITER instead of `ucs::perf_retry_count`. I think better to use `ucs_max(1, ucs::perf_retry_count)`, since it could be configured by a user

Fixed (I just forgot to push).

> 
> also if i'm not mistaken @yosefe asked to move wait_mem testing to `gtest/ucp/test_ucp_perf.cc` file
> 
> > we should put WFE tests in same test fixture and file as ucp perf test, it will eliminate most code duplication

These are  two quire different tests that happen to leverage the same infrastructure. One compares absolute performance of network operations. Other test estimates relative overhead of wait for memory operation which supposed to be no-op. Code duplication is not substantial so this is why I placed it in a separate file.

**Comment by dmitrygx** [comment]
> > > addressing @yosefe comments in #5861.
> > 
> > 
> > @shamisp looks like, still using MAX_ITER instead of `ucs::perf_retry_count`. I think better to use `ucs_max(1, ucs::perf_retry_count)`, since it could be configured by a user
> 
> Fixed (I just forgot to push).

👍 

> > also if i'm not mistaken @yosefe asked to move wait_mem testing to `gtest/ucp/test_ucp_perf.cc` file
> > > we should put WFE tests in same test fixture and file as ucp perf test, it will eliminate most code duplication
> 
> These are two quite different tests that happen to leverage the same infrastructure. One compares absolute performance of network operations. Other test estimates relative overhead of wait for memory operation which supposed to be no-op. Code duplication is not substantial so this is why I placed it in a separate file.

@shamisp if there are any comments, feel free  nd push directly to my branch, so the fixes will appear in the PR

File: test/gtest/common/test_perf.cc [hunk 1]
```diff
@@ -252,9 +252,8 @@ test_perf::test_result test_perf::run_multi_threaded(const test_spec &test, unsi
     return result;
 }
 
-void test_perf::run_test(const test_spec& test, unsigned flags, bool check_perf,
-                         const std::string &tl_name, const std::string &dev_name,
-                         double *perf_value)
+double test_perf::run_test(const test_spec& test, unsigned flags, bool check_perf,
+                         const std::string &tl_name, const std::string &dev_name)
```
Comment (yosefe) [inline test/gtest/common/test_perf.cc:256]: indent

File: test/gtest/common/test_perf.cc [hunk 2]
```diff
@@ -292,20 +291,17 @@ void test_perf::run_test(const test_spec& test, unsigned flags, bool check_perf,
...
+            return value; /* Success */
         } else {
             ucs::safe_sleep(ucs::perf_retry_interval);
         }
     }
 
-     ADD_FAILURE() << "Invalid " << test.title << " performance, expected: " <<
-                      std::setprecision(3) << test.min << ".." << test.max;
+    ADD_FAILURE() << "Invalid " << test.title << " performance, expected: " <<
+        std::setprecision(3) << test.min << ".." << test.max;
```
Comment (yosefe) [inline test/gtest/common/test_perf.cc:304]: indent

File: test/gtest/common/test_perf.h [hunk 3]
```diff
@@ -12,6 +12,10 @@
 #include <tools/perf/api/libperf.h>
 
 
+#define UCP_ARM_PERF_TEST_MULTIPLIER  2
+#define UCT_ARM_PERF_TEST_MULTIPLIER 15
+#define UCT_PERF_TEST_MULTIPLIER      5
```
Comment (yosefe) [inline test/gtest/common/test_perf.h:17]: can we move these back to CC files?

File: test/gtest/common/test_perf.h [hunk 4]
```diff
@@ -35,9 +39,25 @@ class test_perf {
...
+                test.max *= UCT_PERF_TEST_MULTIPLIER;
+                test.min /= UCT_PERF_TEST_MULTIPLIER;
+            }
+        } else {
+            if (ucs_arch_get_cpu_model() == UCS_CPU_MODEL_ARM_AARCH64) {
+                test.max *= UCP_ARM_PERF_TEST_MULTIPLIER;
+                test.min /= UCP_ARM_PERF_TEST_MULTIPLIER;
+            }
+        }
+    }
```
Comment (yosefe) [inline test/gtest/common/test_perf.h:60]: can we move this back to CC file

File: test/gtest/common/test_perf.h (hunk 3 above)
Comment (dmitrygx) [reply test/gtest/common/test_perf.h:17]: do you mean to return back the whole updates?
I just wanted to re-use code for all perf tests. and I guess it makes it simpler

File: test/gtest/common/test_perf.h (hunk 4 above)
Comment (dmitrygx) [reply test/gtest/common/test_perf.h]: moved to test_perf.cc file

File: test/gtest/common/test_perf.h (hunk 3 above)
Comment (dmitrygx) [reply test/gtest/common/test_perf.h]: moved to test_perf.cc file

File: test/gtest/common/test_perf.cc (hunk 1 above)
Comment (dmitrygx) [reply test/gtest/common/test_perf.cc]: fixed

File: test/gtest/common/test_perf.cc (hunk 2 above)
Comment (dmitrygx) [reply test/gtest/common/test_perf.cc]: fixed

**Comment by dmitrygx** [comment]
@shamisp decided to prepare and merge #5942 workaround
can you pls create the new PR and address all comments + fix the original issue? thanks!
closing this

**Comment by shamisp** [comment]
@dmitrygx - no problem but is going to be largely the same PR :)

**Comment by dmitrygx** [comment]
> @dmitrygx - no problem but is going to be largely the same PR :)

@shamisp it is even better, but it mainly consists of your code. my part is just moving code and using `perf_retry_count` value
also, @yosefe commented that it will be good to move WFE to `test_ucp_perf.cc`

**Comment by shamisp** [comment]
I will take a look later today.

**Pull request closed without merging**